lib/bolt/cli.rb

```diff
@@ -379,710 +341,584 @@ def validate(options)
+            client.bundled_content = bundled_content(options)
+            client
+          end
 
-      analytics.screen_view(screen, **screen_view_fields)
+          # TODO: Configure logger with a separate method call?
+          Bolt::Logger.configure(config.log, config.color, config.disable_warnings)
+          Bolt::Logger.stream = config.stream
+          Bolt::Logger.analytics = analytics
+          Bolt::Logger.flush_queue
+
+          executor = Bolt::Executor.new(
+            config.concurrency,
+            analytics,
+            options[:noop],
+            config.modified_concurrency,
+            config.future
+          )
+
+          pal = Bolt::PAL.new(
+            Bolt::Config::Modulepath.new(config.modulepath),
+            config.hiera_config,
+            config.project.resource_types,
+            config.compile_concurrency,
+            config.trusted_external,
+            config.apply_settings,
+            config.project
+          )
+
+          plugins = Bolt::Plugin.setup(config, pal, analytics)
+
+          inventory = Bolt::Inventory.from_config(config, plugins)
+
+          log_outputter = Bolt::Outputter::Logger.new(options[:verbose], config.trace)
+
+          #
+          # FINALIZING SETUP
+          #
+
+          check_gem_install
+          warn_inventory_overrides_cli(config, options)
+          submit_screen_view(analytics, config, inventory, options)
+          process_target_list(plugins.puppetdb_client, @rerun, options)
+
+          # TODO: Fix casing issue in Windows.
+          config.check_path_case('modulepath', config.modulepath)
+
+          if options[:clear_cache] && File.exist?(config.project.plugin_cache_file)
+            FileUtils.rm(config.project.plugin_cache_file)
+          end
 
-      case options[:action]
-      when 'show'
-        case options[:subcommand]
-        when 'task'
-          if options[:object]
-            show_task(options[:object])
+          # TODO: Include logic for handling special cases when subscribing the
+          #       outputter (e.g. `plan run` and `apply`).
+          case command
+          when 'apply', 'lookup'
+            if %w[human rainbow].include?(config.format)
+              executor.subscribe(outputter)
+            end
+          when 'plan'
+            if %w[human rainbow].include?(config.format)
+              executor.subscribe(outputter)
+            else
+              executor.subscribe(outputter, %i[message verbose])
+            end
           else
-            list_tasks
+            executor.subscribe(outputter)
           end
-        when 'plan'
-          if options[:object]
-            show_plan(options[:object])
-          else
-            list_plans
+
+          executor.subscribe(log_outputter)
+
+          # TODO: Figure out where this should really go. It doesn't seem to
+          #       make sense in the application, since the params should already
```

Comment:
Actually, since there's only 2 cases where we do this it seems like they could just go in the `task run` and `plan run` handlers. Similarly the `options[:argument]` for script seems unnecessary.

Reply:
This seems similar to `query` in that it's a CLI implementation detail. From what I understand, `pal.parse_params` converts JSON values to the expected parameter type. Ideally, the application receives parameters as actual objects, not JSON, hence they should be parsed before reaching the application, i.e. params would always be passed as:

```ruby
application.run_plan(..., params: { 'foo' => [1, 2, 3] })
```

and not

```ruby
application.run_plan(..., params: { 'foo' => "[1, 2, 3]" })
```